pkg/controller/change/device/validation.go

```diff
@@ -0,0 +1,87 @@
+// Copyright 2020-present Open Networking Foundation.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package device
+
+import (
+	"fmt"
+	"sort"
+
+	devicechange "github.com/onosproject/onos-config/api/types/change/device"
+	"github.com/onosproject/onos-config/pkg/store"
+	"github.com/onosproject/onos-config/pkg/utils"
+)
+
+func (r *Reconciler) validateDeviceChange(change *devicechange.DeviceChange) error {
```

Comment:
validation should be on the whole configuration plus the DeviceChange - not just on a single DeviceChange on its own

A single DeviceChange might not be valid on its own, and even if it was it does not tell you whether the whole configuration is valid after the DeviceChange would be a applied

Reply:
@SeanCondon Thanks for starting the review. Just to make it clear for myself. Do you mean a network change (i.e. multiple device changes) is not valid if any of device changes is not valid? If that is the case which is similar to what we do in NB by going through all the changes in a loop to validate, I took care of it in Network change controller to fail a network change if validation of device changes is failed. If you are talking about a new logic that should be added to what we have now, please explain it with more details and I will add it.

Reply:
@adibrastegarnia What I meant is that the validation should be applied to:
all previous DeviceChanges for that device + Snapshot for the Device + the new DeviceChange
(as if they were all squashed down to one config).

I think from the code you are applying the validation to only the DeviceChange in hand.

Reply:
@SeanCondon 
Yeah I didn't handle that because I just moved what we have in NB which seems validate just the current request coming from NB to the controller to validate the requested change but what would be the logic (validation function) look like for that kind of validation which should consider all of the previous changes and snapshot?  It sounds like we should retrieve everything build a new json tree using ygot and validate that tree which has the whole configuration including the new change.

Reply:
@adibrastegarnia See line 51 of setconfig.go in existing code
https://github.com/onosproject/onos-config/blob/d0ff6cf04aa47e02244fad1fe9a0398a69de522a/pkg/manager/setconfig.go#L51
It is pulling all the existing changes from store and merges them with the new proposed changes and then does build tree on the whole lot and passes this for validation

Reply:
@SeanCondon 
Thanks for the pointer. I thought that one is returning the last change so I missed it.